resolver/addressable_resolver.go

```diff
@@ -38,16 +38,21 @@ import (
 	"knative.dev/pkg/tracker"
 )
 
+type URIResolverFunc func(ctx context.Context, ref *corev1.ObjectReference) (*apis.URL, error)
+
 // URIResolver resolves Destinations and ObjectReferences into a URI.
 type URIResolver struct {
 	tracker       tracker.Interface
 	listerFactory func(schema.GroupVersionResource) (cache.GenericLister, error)
+	resolvers     []URIResolverFunc
 }
 
 // NewURIResolver constructs a new URIResolver with context and a callback
 // for a given listableType (Listable) passed to the URIResolver's tracker.
-func NewURIResolver(ctx context.Context, callback func(types.NamespacedName)) *URIResolver {
-	ret := &URIResolver{}
+func NewURIResolver(ctx context.Context, callback func(types.NamespacedName), resolvers ...URIResolverFunc) *URIResolver {
```

Comment:
Have you considered using  composition here rather than adding logic to the existing struct? The URIResolverFunc seems like it probably wants to be an interface eventually anyway (once it gets parameterised and has an http client in it), so I wonder whether we could keep NewURIResolver as it is, and then add eg NewLookasideURIResolver(ctx, callback, fallback). That would also make it easier to - for example - opt out of the fallback case or alter the logic for when a resolver does or doesn't fall back.

Reply:
Ugh I just realised we use the struct directly rather than a Resolver interface in loads of places so we'd need to change all of those to use an interface. Bah. Oh well.

Reply:
revolved?

Reply:
Yeah, I think there's a potentially nice refactor here, but it's way more of a can of worms than I thought. Either way I wouldn't block the PR on it 👍